modules/ml/src/main/java/org/apache/ignite/ml/tree/randomforest/data/impurity/GiniHistogram.java

```diff
@@ -0,0 +1,205 @@
+    /** Bucket meta. */
+    private final BucketMeta bucketMeta;
+
+    /** Feature id. */
+    private final int featureId;
+
+    /** Sample id. */
+    private final int sampleId;
+
+    /** Hists of counters for each labels. */
+    private final ArrayList<ObjectHistogram<BootstrappedVector>> hists;
+
+    /** Label mapping to internal representation. */
+    private final Map<Double, Integer> lblMapping;
+
+    /** Bucket ids. */
+    private final Set<Integer> bucketIds;
+
+    /**
+     * Creates an instance of GiniHistogram.
+     *
+     * @param sampleId Sample id.
+     * @param lblMapping Label mapping.
+     * @param bucketMeta Bucket meta.
+     */
+    public GiniHistogram(int sampleId, Map<Double, Integer> lblMapping, BucketMeta bucketMeta) {
+        this.hists = new ArrayList<>(lblMapping.size());
+        this.featureId = bucketMeta.getFeatureMeta().getFeatureId();
+        this.sampleId = sampleId;
+        this.bucketMeta = bucketMeta;
+        this.lblMapping = lblMapping;
+
+        for (int i = 0; i < lblMapping.size(); i++)
+            hists.add(new ObjectHistogram<>(this::bucketMap, this::counterMap));
+
+        this.bucketIds = new TreeSet<>();
+    }
+
+    /** {@inheritDoc} */
+    @Override public void addElement(BootstrappedVector vector) {
+        Integer lblId = lblMapping.get(vector.getLabel());
+        hists.get(lblId).addElement(vector);
+    }
+
+    /** {@inheritDoc} */
+    @Override public Optional<Double> get(Integer bucket) {
+        throw new NotImplementedException();
+    }
+
+    /** {@inheritDoc} */
+    @Override public void addHist(GiniHistogram other) {
+        assert featureId == other.featureId;
+        assert sampleId == other.sampleId;
+
+        bucketIds.addAll(other.bucketIds);
+        for (int i = 0; i < hists.size(); i++)
+            hists.get(i).addHist(other.hists.get(i));
+    }
+
+    /** {@inheritDoc} */
+    @Override public Optional<NodeSplit> findBestSplit() {
+        if(bucketIds.size() < 2)
+            return Optional.empty();
+
+        double bestImpurity = Double.POSITIVE_INFINITY;
+        double bestSplitValue = Double.NEGATIVE_INFINITY;
+        int bestBucketId = -1;
+
+        List<TreeMap<Integer, Double>> distributions = hists.stream().map(ObjectHistogram::computeDistributionFunction)
+            .collect(Collectors.toList());
+        double[] totalCounts = distributions.stream()
+            .mapToDouble(x -> x.isEmpty() ? 0.0 : x.lastEntry().getValue())
+            .toArray();
+
+        Map<Integer, Double> lastLeftValues = new HashMap<>();
+        for (int i = 0; i < lblMapping.size(); i++)
+            lastLeftValues.put(i, 0.0);
+
+        for (Integer bucketId : bucketIds) {
+            double totalToleftCnt = 0;
+            double totalToRightCnt = 0;
+
+            double leftImpurity = 0;
+            double rightImpurity = 0;
+
+            for (int lbId = 0; lbId < lblMapping.size(); lbId++) {
```

Comment:
could you please refactor this algorithm to the high-level composition of procedures/functions to easy understand and support the algorithm in future

Reply:
I added several comments